src/mod_sasl2.erl

```diff
@@ -0,0 +1,442 @@
+success_stanza(SaslAcc, C2SData, MaybeCData) ->
+    Jid = mongoose_c2s:get_jid(C2SData),
+    Inlines = get_acc_sasl2_state(SaslAcc),
+    InlineAnswers = get_inline_responses(Inlines),
+    case MaybeCData of
+        undefined ->
+            AuthorizationId = success_subelement(<<"authorization-identifier">>, jid:to_binary(Jid)),
+            sasl2_ns_stanza(<<"success">>, [AuthorizationId | InlineAnswers]);
+        CData ->
+            AdditionalData = success_subelement(<<"additional-data">>, base64:encode(CData)),
+            AuthorizationId = success_subelement(<<"authorization-identifier">>, jid:to_binary(Jid)),
+            sasl2_ns_stanza(<<"success">>, [AdditionalData, AuthorizationId | InlineAnswers])
+    end.
+
+-spec get_inline_responses([inline_request()]) -> [exml:element()].
+get_inline_responses(Inlines) ->
+    [ Response || {Module, #{status := Status, response := Response}} <- Inlines,
+                  ?MODULE =/= Module,
+                  pending =/= Status,
+                  undefined =/= Response ].
+
+-spec challenge_stanza(binary()) -> exml:element().
+challenge_stanza(ServerOut) ->
+    Challenge = #xmlcdata{content = base64:encode(ServerOut)},
+    sasl2_ns_stanza(<<"challenge">>, [Challenge]).
+
+-spec failure_stanza(binary()) -> exml:element().
+failure_stanza(Reason) ->
+    SaslErrorCode = #xmlel{name = Reason, attrs = [?XMLNS_SASL]},
+    sasl2_ns_stanza(<<"failure">>, [SaslErrorCode]).
+
+-spec sasl2_ns_stanza(binary(), [exml:element() | exml:cdata()]) -> exml:element().
+sasl2_ns_stanza(Name, Children) ->
+    #xmlel{name = Name, attrs = [?XMLNS_SASL2], children = Children}.
+
+-spec success_subelement(binary(), binary()) -> exml:element().
+success_subelement(Name, AuthId) ->
+    #xmlel{name = Name, children = [#xmlcdata{content = AuthId}]}.
+
+%% internal
+-spec get_selected_mech(exml:element()) -> binary().
+get_selected_mech(El) ->
+    exml_query:attr(El, <<"mechanism">>, <<>>).
+
+-spec get_initial_response(exml:element()) -> binary().
+get_initial_response(El) ->
+    base64:decode(exml_query:path(El, [{element, <<"initial-response">>}, cdata], <<>>)).
+
+-spec init_mod_state(not_provided | exml:element()) -> invalid_agent | mod_state().
+init_mod_state(not_provided) ->
+    #{authenticated => false, id => not_provided, software => not_provided, device => not_provided};
+init_mod_state(El) ->
+    MaybeId = exml_query:attr(El, <<"id">>, not_provided),
+    case if_provided_then_is_not_invalid_uuid_v4(MaybeId) of
+        invalid_agent ->
+            invalid_agent;
+        Value ->
+            Software = exml_query:path(El, [{element, <<"software">>}, cdata], not_provided),
+            Device = exml_query:path(El, [{element, <<"device">>}, cdata], not_provided),
+            #{authenticated => false, id => Value, software => Software, device => Device}
+    end.
+
+-spec if_provided_then_is_not_invalid_uuid_v4(not_provided | binary()) ->
+    invalid_agent | uuid:uuid().
```

Comment:
It can also return `not_provided`- see the first clause.

Reply:
Oh dialyzer ☹️